HHH-18724 Take constraint composition type into account when determining nullable state of a Column

wytzevanderploeg opened the pull request:
----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

<!-- Hibernate GitHub Bot issue links start -->
<!-- THIS SECTION IS AUTOMATICALLY GENERATED, ANY MANUAL CHANGES WILL BE LOST -->
https://hibernate.atlassian.net/browse/HHH-18724
<!-- Hibernate GitHub Bot issue links end -->

gavinking commented on hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:
```diff
@@ -263,14 +275,40 @@ private static boolean applyConstraints(
+			Method compositionTypeMethod = descriptor.getClass()
+					.getMethod( "getCompositionType" );
+			Object result = compositionTypeMethod.invoke( descriptor );
+			if ( result != null && "or".equalsIgnoreCase( result.toString() ) ) {
```
Why `equalsIgnoreCase`?

wytzevanderploeg replied:
No particular reason. Just defensive programming I guess.

gavinking commented on hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:
```diff
@@ -263,14 +275,40 @@ private static boolean applyConstraints(
+		// This check assumes that Hibernate Validator is being used providing an
+		// implementation that gives us the composition type that is being used.
+		try {
+			Method compositionTypeMethod = descriptor.getClass()
```
This looks like it's being executed many times.

wytzevanderploeg replied:
I think I can shortcut it a bit more. Should only need to trigger when there is more than one constraint in the composition.

gavinking replied:
I think it should be cached statically.

wytzevanderploeg replied:
Ran code coverage on it based on the tests.
For the early-return part this gives 76 hits, 71 return early, 5 fall through.
Let met know if you consider this acceptable. Otherwise some kind of caching needs to be applied.

wytzevanderploeg replied:
Added static caching.

gavinking reviewed (changes requested):
The pull request should target `main`.

Also see my comments.

gavinking commented on hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:
```diff
@@ -281,16 +298,43 @@ private static boolean applyConstraints(
+			// implementation that gives us the composition type that is being used.
+			try {
+				Method compositionTypeMethod = descriptor.getClass()
+						.getMethod( "getCompositionType" );
```
Wait, is this calling a method of the internal class `ConstraintDescriptorImpl`?

I was expecting this to work be reflecting on `ConstraintComposition` annotations.

wytzevanderploeg replied:
Yes, did not think of this tbh. Makes more sense to reflect the annotations present. Will update accordingly.

wytzevanderploeg commented on hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:
```diff
@@ -328,35 +372,31 @@ private static void applySQLCheck(Column column, String checkConstraint) {
-							);
-						}
+	private static boolean isNotNullDescriptor(ConstraintDescriptor<?> descriptor) {
+		return List.of( NotNull.class, NotEmpty.class, NotBlank.class )
+				.contains( descriptor.getAnnotation().annotationType() );
```
You are completely right. My bad, will revert.

wytzevanderploeg commented:
Fixed edge case. Added a few more tests to ensure better test coverage.
Had to change logic a bit to ensure that initial state of the nullable boolean was set based on the first constraint in the collection.

gavinking commented on hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:
```diff
@@ -328,35 +414,33 @@ private static void applySQLCheck(Column column, String checkConstraint) {
+			//composite should not add not-null on all columns
+			if ( !property.isComposite() ) {
+				for ( Selectable selectable : property.getSelectables() ) {
+					if ( selectable instanceof Column ) {
+						((Column) selectable).setNullable( false );
```
```suggestion
					if ( selectable instanceof Column column ) {
						column.setNullable( false );
```

gavinking commented on hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:
```diff
@@ -257,40 +266,117 @@ private static boolean applyConstraints(
+			return false;
+		}
+
+		Iterator<ConstraintDescriptor<?>> descriptorIterator = constraintDescriptors.iterator();
+		ConstraintDescriptor<?> descriptor = descriptorIterator.next();
+
+		// head
+		boolean hasNotNull = applyConstraint(
+				descriptor,
+				property,
+				propertyDesc,
+				groups,
+				canApplyNotNull,
+				dialect
+		);
+
+		// tail
+		BiFunction<Boolean, Boolean, Boolean> compositionLogic = useOrLogicForComposedConstraint ?
+				Boolean::logicalAnd :
+				Boolean::logicalOr;
+		while (descriptorIterator.hasNext()) {
+			descriptor = descriptorIterator.next();
+			boolean isNotNull = applyConstraint(
+					descriptor,
+					property,
+					propertyDesc,
+					groups,
+					canApplyNotNull,
+					dialect
+			);
+			hasNotNull = compositionLogic.apply( hasNotNull, isNotNull );
+		}
```
All this looks unnecessarily complicated to me. I don't see why we need to use a function reference here, and I don't see why this can't just be written as a `for` loop.

wytzevanderploeg replied:
Fair. Updated.

gavinking commented:
Thanks, looking good now.

beikov commented on hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:
```diff
@@ -72,6 +75,9 @@ class TypeSafeActivator {
 
 	private static final CoreMessageLogger LOG = Logger.getMessageLogger( MethodHandles.lookup(), CoreMessageLogger.class, TypeSafeActivator.class.getName() );
 
+	private static final Map<Class<? extends Annotation>, Boolean>
```
Storing class references to application classes in a static field is unfortunately going to be a problem, unless you clean up the storage properly. Also, concurrent deployments could potentially run into race conditions, so I would suggest you use a `ThreadLocal<Map<Class<? extends Annotation>, Boolean>>` instead and also make sure you clear it in a finally block within the `activate` method.

beikov commented on hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:
```diff
@@ -257,12 +264,17 @@ private static boolean applyConstraints(
 			Dialect dialect) {
-		boolean hasNotNull = false;
+
+		boolean firstItem = true;
+		boolean composedResultHasNotNull = false;
```
```suggestion
		boolean composedResultHasNotNull = !useOrLogicForComposedConstraint;
```

wytzevanderploeg replied:
This is what I initially had but it leads to an edge case when the first property is nullable in a composed constraint of type OR. (See test case I added for the Address.line3 property) In this case the result should be that the column should be considered nullable (nullable OR not-nullable => nullable) This is partly due to the fact that this function is applied recursively for constraints and I did not feel comfortable enough to completely change this. Thus resulting in this head/tails approach.

beikov commented on hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java:
```diff
@@ -276,19 +288,66 @@ private static boolean applyConstraints(
+			}
 
-				hasNotNull = hasNotNull || hasNotNullFromComposingConstraints;
+			if ( firstItem ) {
+				composedResultHasNotNull = hasNotNull;
+				firstItem = false;
+			}
+			else if ( !useOrLogicForComposedConstraint ) {
```
```suggestion
			if ( !useOrLogicForComposedConstraint ) {
```
